# https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html
# https://hbctraining.github.io/DGE_workshop/lessons/08_DGE_LRT.html

#source("/hpc/home/kailasamms/projects/RNASeq/RNASeq_DESeq2_Functions.R")
source("C:/Users/kailasamms/OneDrive - Cedars-Sinai Health System/Documents/GitHub/R-Scripts/RNASeq_DESeq2_Functions.R")
#******************************************************************************#
#                            DEFINE DESEQ2 VARIABLES                           #
#******************************************************************************#

# proj is the directory name containing 
# input files (paste0(proj, "_Metadata.xlsx"), "Read_data.xlsx") or 
# input folders ("count_results" which contain HTSEQ/STAR count output files. 
# The names of files in count_results folder MUST match with Sample_ID column in
# metadata file)

proj <- "EGAD00001007575"
proj <- "EGAD00001003977"
proj <- "RNASeq_Hany_Antigen"
proj <- "RNASeq_Hany_Natural_LOY_cells"
proj <- "RNASeq_Hany_CRISPR_LOY_cells"
proj <- "RNASeq_Boopati_DDR2KO"

# Path to directory containing input files or folders
parent_path <- paste0("C:/Users/KailasammS/Box/Saravana@cedars/05. Bioinformatics/RNASeq/", proj, "/")
results_path <- parent_path

# Choose if data is from human or mice. We will adjust gene names accordingly.
species <- "Homo sapiens"
species <- "Mus musculus"

# Define method used for calculating counts.
# NOTE: A folder named "count_results" containing all the count.txt files
# generated by STAR or HTSEQ MUST be present in parent_path
method <- "HTSEQ"
method <- "STAR"

# Define DESeq2 thresholds
padj.cutoff <- 0.1
lfc.cutoff <- 0     # 0.58 for more strict analysis

# Indicate if you want to plot heatmap, volcano plot, PCA plot
heatmap_plot <- FALSE
volcano_plot <- FALSE
cor_plot  <- FALSE

suffix <- ""

if (proj=="RNASeq_Hany_Antigen"){
  Comparisons <- list(Variable =c("Epitope",  "Growth",     "Condition",         "Condition"),
                      Target   =c("Edited",   "Progressor", "Edited Progressor", "Unedited Progressor"),
                      Reference=c("Unedited", "Regressor",  "Edited Regressor",  "Unedited Regressor"))    
}
if (proj %in% c("RNASeq_Hany_Natural_LOY_cells","RNASeq_Hany_CRISPR_LOY_cells")){
  Comparisons <- list(Variable =c("Condition"),
                      Target   =c("Y_Negative"),
                      Reference=c("Y_Positive"))    
}
if (proj %in% c("RNASeq_Boopati_DDR2KO")){
  Comparisons <- list(Variable =c("Condition"),
                      Target   =c("KO"),
                      Reference=c("Control"))    
}

#******************************************************************************#
#                             READ DATA & META DATA                            #                    
#******************************************************************************#

# read_data MUST be xlsx file
# read_data MUST have "SYMBOL" column with gene names without any duplication
read_data <- openxlsx::read.xlsx(xlsxFile=paste0(parent_path, "Read_data.xlsx"))

# meta_data MUST be xlsx file
# meta_data MUST have "Sample_ID" column with sample names without any duplication
# meta_data MUST have "Batch" column with batch info
# Example: If samples were isolated on different days & prepared using different
# kits, "Batch" column must have values like "1_Ribo", "1_Poly-A", "2_Ribo", etc
# NOTE: Make sure there are no white spaces in the Target and Reference columns
# in excel file. R will change any white space (" ") to dot ("."). So, replace 
# white space (" ") with underscore ("_") before importing into R.
meta_data <- openxlsx::read.xlsx(xlsxFile=paste0(parent_path, proj, "_Metadata.xlsx"))

# If "Read_data.xlsx" is not prepared yet and counts are stored in count files
# read_data <- compile_raw_counts(parent_path, method)

#******************************************************************************#
#                    IMPORT DATA AND RUN THE DESEQ2 PIPLEINE                   #                    
#******************************************************************************#

annotations <- get_annotations(species)
meta_data <- prep_metadata(meta_data, Variable)
read_data <- prep_readdata(read_data, meta_data)
l <- check_data(read_data, meta_data)
meta_data <- l[[2]]
read_data <- l[[1]]

# Perform DEG analysis for each Target:Reference pair in Comparisons variable

# NOTE: The normalized counts you get from sva_dds and DESeq2 dds are NOT batch 
# corrected and they will be identical. Both DESeq2() and svaseq() ONLY model
# for batch factors/surrogate variables in the design formula, they DO NOT 
# modify counts like combatseq. The normalized counts you get from combatseq
# are batch corrected. 

# NOTE: Lowly expressed genes are removed before finding surrogate variables
# in svaseq(). As a result, the number of genes in normalized counts excel file
# is lower than DeSeq2.

for (n in 1:length(Comparisons$Variable)){
  
  # This generates a new column "id" that has info on samples being comparared
  meta_data_comp <- meta_data %>%
    dplyr::mutate(id=get(Comparisons$Variable[n]))
  
  #combat_corrected_read_data <- combatseq_batch(read_data, meta_data_comp)
  #sva_dds <- svaseq_batch(read_data, meta_data_comp)
  
  # Perform DESeq2() using in-built batch modelling
  approach <- "DESeq2_modelled"
  if (length(unique(meta_data_comp$Batch)) > 1){
    dds <- DESeq2::DESeqDataSetFromMatrix(countData=read_data,
                                          colData=meta_data_comp, 
                                          design=~ Batch+id)
  } else {
    dds <- DESeq2::DESeqDataSetFromMatrix(countData=read_data,
                                          colData=meta_data_comp, 
                                          design=~ id)
  }
  dds <- run_deseq2(dds, meta_data_comp, annotations, Comparisons, n, approach, prefix, results_path)
  deseq2_norm_counts(dds, annotations, approach, suffix) # batch corrected if you more than 1 batch
  plot_qc(dds, meta_data_comp, approach, suffix)
  
  # Perform DESeq2() using combat corrected read counts
  if (!identical(read_data, combat_corrected_read_data)){
    approach <- "combat_corrected"
    combat_dds <- DESeq2::DESeqDataSetFromMatrix(countData=combat_corrected_read_data,
                                                 colData=meta_data_comp, 
                                                 design=~ id)
    combat_dds <- run_deseq2(combat_dds, meta_data_comp, annotations, Comparisons, n, lfc.cutoff, padj.cutoff, suffix)
    combatseq_norm_counts(combat_dds, annotations, approach, suffix)  #combat batch corrected
    plot_qc(combat_dds, meta_data_comp, approach, suffix)
  }
  
  # Perform DESeq2() using sva modelled surrogate variables SV1 and SV2
  approach <- "sva_modelled"
  sva_dds <- run_deseq2(sva_dds, meta_data_comp, annotations, Comparisons, n, lfc.cutoff, padj.cutoff, approach, suffix)
  # calc_norm_counts(sva_dds, annotations, approach, suffix)   # uncorrected
  svaseq_norm_counts(sva_dds, annotations, approach, suffix)   # sva seq batch corrected
  plot_qc(sva_dds, meta_data_comp, approach, suffix)
}

#*************************************END**************************************#

# Define Target and Reference base don project
if (proj == "TCGA"){
  Variable <- "Sex"
  Comparisons <- list(Target=c("Male"),
                      Reference=c("Female"))
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj == "IMVigor210"){
  Variable <- "binaryResponse"
  Comparisons <- list(Target=c("SD/PD"),
                      Reference=c("CR/PR"))
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj =="Mukta_GSE95097"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("8hr", "4hr"),
                      Reference=c("0hr", "0hr"))
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj == "Boopati_MB49Y-_DDR2KO"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Ddr2KO"),
                      Reference=c("Control"))
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj == "Boopati_MB49Sigma_DDR2KD3"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Ddr2KD"),
                      Reference=c("Control"))
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="Mukta_NA13_MB49_CDH12OE"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("OE"),
                      Reference=c("Control"))
  Variable2 <- "Cell_line"
  Variable2_value <- "NA13"  #MB49"
}
if (proj=="Hany_KO"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Kdm5d_KO", "Uty_KO"), 
                      Reference=c("scr_KO", "scr_KO"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="Hany_YKO"){
  Variable <- "Condition"
  Comparisons <- list(Target=   c("YKO_centromere"), 
                      Reference=c("scr_KO_centromere"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="Hany_Y"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Y_neg"),
                      Reference=c("Y_pos"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="Hany_Y_Tumor"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Y_pos_Tumor", "Hany_Tumor"),
                      Reference=c("Y_neg_Tumor", "Sigma_Tumor"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="Hany_OE"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Kdm5d_OE", "Uty_OE"),
                      Reference=c("scr_OE", "scr_OE"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="Hany_OE_Tumor"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Kdm5d_OE_Tumor", "Uty_OE_Tumor"),
                      Reference=c("scr_OE_Tumor", "scr_OE_Tumor"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="TRAMP_GSE79756"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("PAX8-NFE2L2 fusion"),
                      Reference=c("Empty vector"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="TCGA"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Male"),
                      Reference=c("Female"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="Prince_DepMap"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Ypos", "Female"),
                      Reference=c("Yneg", "Yneg"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="scRNASeq_BBN_Nude"){
  Variable <- "Sex"
  Comparisons <- list(Target=c("Male"),
                      Reference=c("Female"))    
  Variable2 <- NULL
  Variable2_value <- NULL
}
if (proj=="scRNASeq"){
  Variable <- "Ystatus"
  Comparisons <- list(Target=c("Yneg"),
                      Reference=c("Ypos"))    
  Variable2 <- "Condition"
  Variable2_value <- "Tumor"
}
if (proj=="Jinfen_CRISPR"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("Day7_2D", "Day14_2D", "Day7_3D", "Day14_3D"),
                      Reference=c("Day0_2D", "Day0_2D", "Day0_3D", "Day0_3D"))    
  Variable2 <- ""
  Variable2_value <- ""
}
if (proj=="PRJNA587619"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("BBN"),
                      Reference=c("Control"))    
  Variable2 <- ""
  Variable2_value <- ""
}
if (proj=="GSE75192"){
  Variable <- "Condition"
  Comparisons <- list(Target=c("09m_Brain", "15m_Brain", "24m_Brain", "30m_Brain", 
                               "09m_Liver", "15m_Liver", "24m_Liver", "30m_Liver",
                               "09m_Skin",  "15m_Skin",  "24m_Skin",  "30m_Skin",
                               "09m_Whole Blood", "15m_Whole Blood", "24m_Whole Blood", "30m_Whole Blood"),
                      Reference=c("02m_Brain", "02m_Brain", "02m_Brain", "02m_Brain",
                                  "02m_Liver", "02m_Liver", "02m_Liver", "02m_Liver",
                                  "02m_Skin",  "02m_Skin",  "02m_Skin", "02m_Skin",
                                  "02m_Whole Blood", "02m_Whole Blood", "02m_Whole Blood","02m_Whole Blood"))    
  Variable2 <- ""
  Variable2_value <- ""
}
if (proj=="EGAD00001007575"){
  Variable <- "Sex"
  Comparisons <- list(Target=c("Male"),
                      Reference=c("Female"))    
  Variable2 <- ""
  Variable2_value <- ""
}
if (proj=="EGAD00001003977"){
  Variable <- "Sex"
  Comparisons <- list(Target=c("Male"),
                      Reference=c("Female"))    
  Variable2 <- ""
  Variable2_value <- ""
}

#***************************DEFINE HEATMAP VARIABLES***************************#

# Define if data is log transformed already
already_log <- FALSE

# Define if data is scaled already
already_scaled <- FALSE

# Define if you want to perform unsupervised row and column clustering
# NOTE: If set to FALSE, samples (columns) and genes(rows) will be arranged 
# in alphabetical order (default) in heatmap. If you want to arranged in 
# specific order, define below.
row_clustering <- TRUE    # Usually TRUE
col_clustering <- TRUE    # Usually FALSE

# Define if you want genes or samples to be arranged in alphabetical order
# NOTE: If set to FALSE, write the plot_genes in order you want in heatmap
# NOTE: If row_clustering==TRUE, then row_clustering_alphabetical is irrelevant
row_clustering_alphabetical <- TRUE
col_clustering_alphabetical <- FALSE

# List annotations you want on heatmap
# NOTE: anno_columns MUST match one of the column names in metadata
anno_columns <- c("Condition")

# Define colors for heatmap
my_palette <- colorRampPalette(rev(brewer.pal(n=11, name ="RdYlBu")))(200)
#my_palette <- viridis(200)

# Dimensions of each box in heatmap
cell_width <- NA
cell_height <- NA

# #*******************************DIAGNOSTIC TESTS*******************************#

# # (i) To view counts of specific gene across samples
# plotCounts(dds, gene=which.min(res$padj), intgroup=Variable)           # gene with lowest padj
# plotCounts(dds, gene=which.min(res$log2FoldChange), intgroup=Variable) # gene with lowest log2FC
# plotCounts(dds, gene=which.max(res$log2FoldChange), intgroup=Variable) # gene with highest log2FC
# plotCounts(dds, gene="ENSMUSG00000030598", intgroup=Variable)          # a specific gene
# 

# # To identify the genes interactively, run the 2 lines below. 
# # Then click on multiple dots and click Finish. A list of genes will be displayed 
# # idx <- identify(res$baseMean, res$log2FoldChange)
# # rownames(res)[idx]
# 
# # (iii) Plot PCA using rld or vst
# # Variable is defined at beginning of DESeq2. So, we set intgroup=Variable.
# rld <- rlog(object=dds,
#             blind=TRUE,
#             # intercept=,
#             # betaPriorVar=,
#             fitType="parametric")
# 
# vst <- varianceStabilizingTransformation(object=dds,
#                                          blind=TRUE,
#                                          fitType="parametric")
# 
# for (process in c("rld", "vst")){
#   
#   plotPCA(object=get(process),
#           intgroup=Variable,
#           ntop=500,
#           returnData=FALSE) +
#     geom_text(aes(label=name), nudge_x=2, nudge_y=2)
#   
#   ggplot2::ggsave(filename=paste0("Diagnostic_PCA_plot_using_", process, "_", celltype, ".pdf"),
#                   plot=last_plot(),
#                   device="pdf",
#                   path=diagnostics_path,
#                   scale=1,
#                   #width=8.5,
#                   #height=11,
#                   units=c("in"),
#                   dpi=300,
#                   limitsize=TRUE,
#                   bg="white")
# }
# 
# # (iv) Hierarchical clustering of samples using rld or vst
# rld_mat <- DESeq2::assay(rld)     # extract the matrix from a DESeq2 object
# rld_cor <- cor(x=rld_mat,       # compute pairwise correlation values
#                y=NULL,
#                use="everything",
#                method="pearson") 
# 
# # Check the output of cor(), make note of the rownames and colnames
# head(rld_cor)
# 
# vst_mat <- assay(vst)  
# vst_cor <- cor(vst_mat) 
# 
# # Check the output of cor(), make note of the rownames and colnames
# head(vst_cor)    
# 
# for (process in c("rld", "vst")){
#   
#   pheatmap::pheatmap(mat=get(paste0(process, "_cor")),
#                      color=colorRampPalette(rev(brewer.pal(n=11, name ="RdYlBu")))(100),
#                      breaks=NA, 
#                      border_color="white", #"grey60",
#                      cellwidth=NA, 
#                      cellheight=NA, 
#                      scale="none",   
#                      cluster_rows=TRUE,   #cluster the rows
#                      cluster_cols=TRUE,   #cluster the columns
#                      clustering_distance_rows="euclidean",
#                      clustering_distance_cols="euclidean",
#                      clustering_method="complete",
#                      legend=TRUE, 
#                      legend_breaks=NA,
#                      legend_labels=NA, 
#                      #annotation_row=,  
#                      #annotation_col=, 
#                      annotation_colors=dplyr::if_else(nrow(col_annotation)+nrow(row_annotation) > 0, ann_colors, NA),
#                      annotation_legend=TRUE,
#                      annotation_names_row=TRUE,
#                      annotation_names_col=TRUE,
#                      show_rownames=dplyr::if_else(nrow(mat)<80, TRUE, FALSE, missing=NULL), 
#                      show_colnames=dplyr::if_else(ncol(mat)<30, TRUE, FALSE, missing=NULL),
#                      fontsize=8, 
#                      fontsize_row=8, 
#                      fontsize_col=8,
#                      angle_col=c("270", "0", "45", "90", "315"),
#                      fontsize_number=0.8*fontsize, 
#                      #labels_row=display_row,
#                      #labels_col=display_col,
#                      filename=paste0(diagnostics_path, "Diagnostic_Correlation_Heatmap_using_", process, "_", celltype, ".pdf"))
# }
# 
# # (v) Checking if mean < variance (for NB model) or Mean=Variance (for Poisson model). 
# # Each point is a gene denoted by (x,y) where 
# # x=mean_count of gene across all samples & y=variance_count of gene across all samples
# 
# mean_counts <- apply(read_data[,], 1, mean)
# variance_counts <- apply(read_data[,], 1, var)
# df <- data.frame(mean_counts, variance_counts)
# ggplot(df) +
#   geom_point(aes(x=mean_counts, y=variance_counts)) +
#   geom_line(aes(x=mean_counts, y=mean_counts, color="red")) +
#   scale_y_log10() +
#   scale_x_log10()
# 
# ggplot2::ggsave(filename=paste0("Diagnostic_Scatter_plot_", celltype, ".pdf"),
#                 plot=last_plot(),
#                 device="pdf",
#                 path=diagnostics_path,
#                 scale=1,
#                 #width=8.5,
#                 #height=11,
#                 units=c("in"),
#                 dpi=300,
#                 limitsize=TRUE,
#                 bg="white")
# 

# 
# # # (vii) Plot p-value histogram
# # hist(res$pvalue, col="lightblue", breaks=20)

# # # (ix) Plot a histogram for one of the samples to see how the counts are distributed. Adjust xlim if needed
# # ggplot(read_data, aes(x=results.S10_R1_trimmed.fastq.gz.csv)) +
# #   geom_histogram(stat="bin", bins=200) +
# #   xlim(-5,1000) +
# #   xlab("Raw expression counts") +
# #   ylab("Number of genes") +
# #   scale_color_brewer(palette="Dark2")
# # 
# # # (x) Extract counts, size factors, etc.. from dds
# # dds <- estimateSizeFactors(dds)         #redundant if you already ran dds <- DESeq(dds)
# # counts <- counts(dds)                   #counts[i,j]=raw_count of gene i in sample j
# # sizefactors <- sizeFactors(dds)         #sizefactors[j]=median (counts[,j]/geometric_mean(counts[i,]))
# # colSums(counts(dds))                    #Total number of raw counts per sample
# # colSums(counts(dds, normalized=TRUE))   #Total number of normalized counts per sample

#******************************************************************************#
#                         TO SUBSET DATA OR NOT!!!                          
#******************************************************************************#

# You have 3 groups A, B & C but want to perform DEG between group A and B only.
# Should you exclude group C samples and perform DEG analysis??
# NOTE: Subsetting data affects (i) sizeFactors (ii) Dispersion estimates and
# hence the final DEGs

# The links below explain when data needs to be subset.
# https://support.bioconductor.org/p/108471/
# https://support.bioconductor.org/p/81038/
# https://support.bioconductor.org/p/9156036/
# https://support.bioconductor.org/p/69341/

# The authors of DESeq2 and EdgeR recommend NOT to subset meta_data & 
# read_data but rather use contrasts to get DEGs between groups
# You should subset your data ONLY when one of the groups has large variance 
# (like group C) in link below
# https://master.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#if-i-have-multiple-groups-should-i-run-all-together-or-split-into-pairs-of-groups

# Dispersion of gene X in group A ~ Variance of gene X in group A/mean of gene X group A
# Dispersion of gene X in group B ~ Variance of gene X in group B/mean of gene X group B
# Dispersion of gene X in group C ~ Variance of gene X in group C/mean of gene X group C
# Dispersion of gene X for experiment ~ (Dispersion of gene X in groups A, B, C)
# Samples within Group A and B have low variance in overall expression profile 
# (clustered closely in PCA plot) but samples in group C have different
# expression profile. So, the final dispersion estimates of most genes will be 
# affected a lot by group C. In such a scenario, if we are only doing DEG 
# comparison between groups A & B, then we can subset only group A and B samples
# before doing DESeq2. 
# SO, CHECK PCA PLOT FOR EVERY EXPERIMENT TO DETERMINE IF DATA NEED TO BE SUBSET.

# meta_data <- meta %>% dplyr::filter(get(Variable) %in% c(Comparisons$Target[n], Comparisons$Reference[n]))
# corrected_read_data <- read %>% dplyr::select(rownames(meta_data))
# NOTE: sizefactors MUST be calculated ONLY using samples being compared for 
# differential expression. So, make sure read_data and meta_data ONLY have
# samples being compared for differential expression.

# dds <- DESeq2::DESeqDataSetFromMatrix(countData=corrected_read_data, colData=meta_data, design=~ Tissue + Age)
# dds <- DESeq(dds, test="LRT", reduced=~ Tissue)
# res <- DESeq2::results(object=dds)

# Study effect of Treatment ignoring effect of Sex
#dds <- DESeqDataSetFromMatrix(countData=corrected_read_data, colData=meta_data, design=~ Treatment)

# Study effect of Treatment after removing effect of Sex (assumes equal effect of Sex on different treatments)
#dds <- DESeqDataSetFromMatrix(countData=corrected_read_data, colData=meta_data, design=~ Sex + Treatment)

# Study effect of Sex on treatment?
# Study effect of Treatment after removing effect of Sex (assumes different effect of Sex on different treatments)
#dds <- DESeqDataSetFromMatrix(countData=corrected_read_data, colData=meta_data, design=~ Sex + Treatment + Sex:Treatment)


# TO DO
# (i) add script to subset data and perform analysis
# (ii) Add script for different design
# (iii) how to handle single sample with group C having high variability?
# (iv) add more QC plots from deseq2 vignette
# (v) Fix script for heatmap, volcano and correlation plot
# (vi) A more intuitive way to link proj with variable and comparisons, target, refence

