# This script is used to compare cluster markers from multiple NGS datasets
# from time to time to generate a master list of markers that can be used to 
# annotate single cell datasets

# We will find markers using 2 approaches
# (i) Standard clusters: clusters generated using harmony at resolution of 0.4
# (ii) Custom clusters: manual grouping of clusters based on (i). Sometimes, 
# epithelial cluster has multiple subclusters generated by harmony. So, we merge
# such subclusters in order to identify macro-level markers

#******************************************************************************#
#                                STEP 1: FIND MARKERS                          #
#******************************************************************************#

# Create dataframe to store output of FindMarkers() from standard approach
df_standard <- data.frame(cluster = "", 
                          gene = "",
                          avg_log2FC = 0,
                          p_val_adj = 0,
                          pct.1 = 0.0, 
                          pct.2 = 0.0,
                          ratio = 0.0)

# Create dataframe to store output of FindMarkers() from standard approach
df_custom <- df_standard

# Iterate through all dataset, find markers using standard & custom approaches
for (proj in c("scRNASeq_BBN_C57B6", "scRNASeq_BBN_Rag", "scRNASeq_Simon", 
               "scRNASeq_Jinfen", "scRNASeq_Chen")){
  
  seurat_results <- paste0("/hpc/home/kailasamms/scratch/", proj, "/results_seurat/")
  res <- 0.4
  reduc <- "harmony"
  celltype <- NULL
  
  # Load the integrated seurat object
  integrated_seurat <- readRDS(paste0(seurat_results, "integrated_seurat_snn", 
                                      dplyr::if_else(is.null(celltype), ".rds", paste0("_", celltype, ".rds"))))
  
  # Change active.ident to major clusters we annotated
  DefaultAssay(integrated_seurat) <- "RNA"
  idents <- paste0("cluster.", res, ".", base::tolower(reduc))
  Idents(object = integrated_seurat) <- idents
  
  # Find ALL markers
  all_markers_standard <- Seurat::FindAllMarkers(object=integrated_seurat,
                                                 assay = "RNA",
                                                 features = NULL,
                                                 logfc.threshold = 0.25,
                                                 test.use = "wilcox",
                                                 slot = "data",
                                                 min.pct = 0.1,
                                                 min.diff.pct = 0.1,
                                                 node = NULL,
                                                 verbose = TRUE,
                                                 only.pos = TRUE,
                                                 max.cells.per.ident = Inf,
                                                 random.seed = 1,
                                                 latent.vars = NULL,
                                                 min.cells.feature = 3,
                                                 min.cells.group = 1,
                                                 pseudocount.use = 1,
                                                 mean.fxn = NULL,
                                                 fc.name = NULL,
                                                 base = 2,
                                                 return.thresh = 0.01,
                                                 densify = FALSE)
  
  all_markers_standard <- all_markers_standard %>% 
    dplyr::mutate(pct.1 = dplyr::if_else(pct.1 == 0, 0.001, pct.1),
                  pct.2 = dplyr::if_else(pct.2 == 0, 0.001, pct.2),
                  ratio = pct.1/pct.2,
                  cluster = paste0(proj, "_", cluster)) %>%
    dplyr::relocate(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, ratio)
  
  
  # Create custom grouping of major clusters
  if (proj == "scRNASeq_BBN_C57B6"){
    clusters <- list("cluster_1" = c(0,22),
                     "cluster_2" = c(3,16),
                     "cluster_3" = c(5,12),
                     "cluster_4" = c(1,6,9),
                     "cluster_5" = c(2,4,10,11,13),
                     "cluster_6" = c(7),
                     "cluster_7" = c(8),
                     "cluster_8" = c(14),
                     "cluster_9" = c(15),
                     "cluster_10" = c(17),
                     "cluster_11" = c(18),
                     "cluster_12" = c(19),
                     "cluster_13" = c(20),
                     "cluster_14" = c(21),
                     "cluster_15" = c(23),
                     "cluster_16" = c(24))
  }
  
  if (proj == "scRNASeq_BBN_Rag"){
    clusters <- list("cluster_1" = c(0,4,7,8,9,10,12,13,15,17,18,20,26,27,31),
                     "cluster_2" = c(3,16,22),
                     "cluster_3" = c(5,24),
                     "cluster_4" = c(1,11,23,25,28),
                     "cluster_5" = c(2,6,21,29,32,33),
                     "cluster_6" = c(14),
                     "cluster_7" = c(19),
                     "cluster_8" = c(30))
  }
  
  if (proj == "scRNASeq_Simon"){
    clusters <- list("cluster_1" = c(0,2,4,10,11,17,19),
                     "cluster_2" = c(3,5,27),
                     "cluster_3" = c(1,16,25,28),
                     "cluster_4" = c(7,13,15),
                     "cluster_5" = c(8,14,20),
                     "cluster_6" = c(6),
                     "cluster_7" = c(9),
                     "cluster_8" = c(12),
                     "cluster_9" = c(18),
                     "cluster_10" = c(21),
                     "cluster_11" = c(22),
                     "cluster_12" = c(23),
                     "cluster_13" = c(24),
                     "cluster_14" = c(26),
                     "cluster_15" = c(29))
  }
  
  if (proj == "scRNASeq_Jinfen"){
    clusters <- list("cluster_1" = c(0,6),
                     "cluster_2" = c(1,3,8),
                     "cluster_3" = c(4),
                     "cluster_4" = c(5),
                     "cluster_5" = c(9),
                     "cluster_6" = c(2),
                     "cluster_7" = c(7),
                     "cluster_8" = c(10),
                     "cluster_9" = c(11),
                     "cluster_10" = c(12),
                     "cluster_11" = c(13),
                     "cluster_12" = c(14),
                     "cluster_13" = c(15),
                     "cluster_14" = c(16))
  }
  
  if (proj == "scRNASeq_Chen"){
    clusters <- list("cluster_1" = c(0,1,4,12,13,22),
                     "cluster_2" = c(2,3,17),
                     "cluster_3" = c(9,15),
                     "cluster_4" = c(16,28),
                     "cluster_5" = c(5),
                     "cluster_6" = c(6),                     
                     "cluster_7" = c(7),
                     "cluster_8" = c(8),
                     "cluster_9" = c(10),
                     "cluster_10" = c(11),
                     "cluster_11" = c(14),
                     "cluster_12" = c(18),
                     "cluster_13" = c(19),
                     "cluster_14" = c(20),
                     "cluster_15" = c(21),
                     "cluster_16" = c(23),
                     "cluster_17" = c(24),
                     "cluster_18" = c(25),
                     "cluster_19" = c(26),
                     "cluster_20" = c(27),
                     "cluster_21" = c(29))
  }
  
  integ <- annotate_data(res, reduc, celltype, clusters)
  
  DefaultAssay(integ) <- "RNA"
  Idents(integ) <- "cell_type"
  
  all_markers_custom <- Seurat::FindAllMarkers(object=integ,
                                               assay = "RNA",
                                               features = NULL,
                                               logfc.threshold = 0.25,
                                               test.use = "wilcox",
                                               slot = "data",
                                               min.pct = 0.1,
                                               min.diff.pct = 0.1,
                                               node = NULL,
                                               verbose = TRUE,
                                               only.pos = TRUE,
                                               max.cells.per.ident = Inf,
                                               random.seed = 1,
                                               latent.vars = NULL,
                                               min.cells.feature = 3,
                                               min.cells.group = 1,
                                               pseudocount.use = 1,
                                               mean.fxn = NULL,
                                               fc.name = NULL,
                                               base = 2,
                                               return.thresh = 0.01,
                                               densify = FALSE)
  
  all_markers_custom <- all_markers_custom %>% 
    dplyr::mutate(pct.1 = dplyr::if_else(pct.1 == 0, 0.001, pct.1),
                  pct.2 = dplyr::if_else(pct.2 == 0, 0.001, pct.2),
                  ratio = pct.1/pct.2,
                  cluster = paste0(proj, "_", cluster)) %>%
    dplyr::relocate(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, ratio)
  
  df_standard <- dplyr::bind_rows(df_standard, all_markers_standard)
  df_custom <- dplyr::bind_rows(df_custom, all_markers_custom)
}


#******************************************************************************#
#                     STEP 2: FILTER OUT BAD MARKERS                           #
#******************************************************************************#

# Filter out weak/non-specific markers & identify top 100 markers for every cluster
df_standard <- df_standard %>%
  dplyr::filter(avg_log2FC > 0.58, # custom expression cutoff 
                p_val_adj < 0.05,  # standard cutoff
                pct.1 > pct.2,     # standard cutoff
                pct.1 > 0.4) %>%   # custom percent cutoff
  dplyr::group_by(cluster) %>%
  dplyr::add_count(cluster) %>%
  dplyr::filter(n >= 10) %>%
  dplyr::arrange(desc(avg_log2FC)) %>%  #desc(ratio)
  dplyr::slice_head(n=100) %>%
  ungroup()

df_custom <- df_custom %>%
  dplyr::filter(avg_log2FC > 0.58, # custom expression cutoff 
                p_val_adj < 0.05,  # standard cutoff
                pct.1 > pct.2,     # standard cutoff
                pct.1 > 0.4) %>%   # custom percent cutoff
  dplyr::group_by(cluster) %>%
  dplyr::add_count(cluster) %>%
  dplyr::filter(n >= 10) %>%
  dplyr::arrange(desc(avg_log2FC)) %>%  #desc(ratio)
  dplyr::slice_head(n=100) %>%
  ungroup()

#******************************************************************************#
#            STEP 3: IDENTIFY SIMILAR CLUSTERS ACROSS DATASETS                 #
#******************************************************************************#

# Create similarity matrix indicating number of common genes between any 2 
# clusters across datasets for standard approach
sim_matrix <- matrix(data=0, 
                     nrow= length(unique(df_standard$cluster)), 
                     ncol=length(unique(df_standard$cluster)))
colnames(sim_matrix) <- unique(df_standard$cluster)
rownames(sim_matrix) <- unique(df_standard$cluster)

for (i in 1:ncol(sim_matrix)){
  for (j in 1:nrow(sim_matrix)){
    sim_matrix[i,j] <- length(intersect(tolower(df_standard %>% 
                                                  dplyr::filter(cluster == colnames(sim_matrix)[i]) %>%
                                                  dplyr::select(gene) %>%
                                                  unlist(use.names = FALSE)),
                                        tolower(df_standard %>% 
                                                  dplyr::filter(cluster == rownames(sim_matrix)[j]) %>%
                                                  dplyr::select(gene) %>%
                                                  unlist(use.names = FALSE))))
  }
}

# Group similar clusters together
mat <- sim_matrix
rowclust <- stats::hclust(d = dist(mat))
reordered <- mat[rowclust$order,]
colclust <- stats::hclust(d = dist(t(mat)))
reordered_standard <- reordered[, colclust$order]

# Create similarity matrix indicating number of common genes between any 2 
# clusters across datasets for custom approach
sim_matrix <- matrix(data=0, 
                     nrow= length(unique(df_custom$cluster)), 
                     ncol=length(unique(df_custom$cluster)))
colnames(sim_matrix) <- unique(df_custom$cluster)
rownames(sim_matrix) <- unique(df_custom$cluster)

for (i in 1:ncol(sim_matrix)){
  for (j in 1:nrow(sim_matrix)){
    sim_matrix[i,j] <- length(intersect(tolower(df_custom %>% 
                                                  dplyr::filter(cluster == colnames(sim_matrix)[i]) %>%
                                                  dplyr::select(gene) %>%
                                                  unlist(use.names = FALSE)),
                                        tolower(df_custom %>% 
                                                  dplyr::filter(cluster == rownames(sim_matrix)[j]) %>%
                                                  dplyr::select(gene) %>%
                                                  unlist(use.names = FALSE))))
  }
}

# Group similar clusters together
mat <- sim_matrix
rowclust <- stats::hclust(d = dist(mat))
reordered <- mat[rowclust$order,]
colclust <- stats::hclust(d = dist(t(mat)))
reordered_custom <- reordered[, colclust$order]
  
# Save the clustered similarity matrix
wb <- openxlsx::createWorkbook()
openxlsx::addWorksheet(wb, sheetName = "Heatmap_standard")
openxlsx::writeData(wb, sheet = "Heatmap_standard", x = t(reordered_standard), rowNames = TRUE)
openxlsx::addWorksheet(wb, sheetName = "Heatmap_custom")
openxlsx::writeData(wb, sheet = "Heatmap_custom", x = t(reordered_custom), rowNames = TRUE)
openxlsx::saveWorkbook(wb = wb, file = "Compiled_Markers.xlsx", overwrite = TRUE)

#******************************************************************************#
#   STEP 4: IDENTIFY COMMOM GENES AMONG SIMILAR CLUSTERS ACROSS DATASETS       #
#******************************************************************************#

# Look at heatmap matrix excel file as well as UMAP at resolution 0.4 for each
# dataset. Identify similar clusters across all experiments as demonstrated in
# the pptx.

l1 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(2),
           "scRNASeq_BBN_Rag" = c(8,9,12,15,17,18,26),
           "scRNASeq_Chen" = c(0,13,23,25,1,4,22),
           "scRNASeq_Simon" = c(0,13))

l2 <- list("scRNASeq_Jinfen" = c(10),
           "scRNASeq_BBN_C57B6" = c(4),
           "scRNASeq_BBN_Rag" = c(0),
           "scRNASeq_Chen" = c(),
           "scRNASeq_Simon" = c())

l3 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(18),
           "scRNASeq_BBN_Rag" = c(30),
           "scRNASeq_Chen" = c(),
           "scRNASeq_Simon" = c())

l4 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(7),
           "scRNASeq_BBN_Rag" = c(5,24),
           "scRNASeq_Chen" = c(5,6,21),
           "scRNASeq_Simon" = c(12))

l5 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(),
           "scRNASeq_BBN_Rag" = c(),
           "scRNASeq_Chen" = c(14),
           "scRNASeq_Simon" = c(21))

l6 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(23),
           "scRNASeq_BBN_Rag" = c(),
           "scRNASeq_Chen" = c(28),
           "scRNASeq_Simon" = c())

l7 <- list("scRNASeq_Jinfen" = c(3),
           "scRNASeq_BBN_C57B6" = c(),
           "scRNASeq_BBN_Rag" = c(),
           "scRNASeq_Chen" = c(12),
           "scRNASeq_Simon" = c())

l8 <- list("scRNASeq_Jinfen" = c(14),
           "scRNASeq_BBN_C57B6" = c(8),
           "scRNASeq_BBN_Rag" = c(),
           "scRNASeq_Chen" = c(11),
           "scRNASeq_Simon" = c())

l9 <- list("scRNASeq_Jinfen" = c(4,15),
           "scRNASeq_BBN_C57B6" = c(3),
           "scRNASeq_BBN_Rag" = c(),
           "scRNASeq_Chen" = c(2,3,17),
           "scRNASeq_Simon" = c(6))

l10 <- list("scRNASeq_Jinfen" = c(1,8,9),
            "scRNASeq_BBN_C57B6" = c(16),
            "scRNASeq_BBN_Rag" = c(),
            "scRNASeq_Chen" = c(),
            "scRNASeq_Simon" = c())

l11 <- list("scRNASeq_Jinfen" = c(2,12),
            "scRNASeq_BBN_C57B6" = c(5,12),
            "scRNASeq_BBN_Rag" = c(2,6),
            "scRNASeq_Chen" = c(9),
            "scRNASeq_Simon" = c())

l12 <- list("scRNASeq_Jinfen" = c(11,16),
            "scRNASeq_BBN_C57B6" = c(17),
            "scRNASeq_BBN_Rag" = c(),
            "scRNASeq_Chen" = c(),
            "scRNASeq_Simon" = c())

l13 <- list("scRNASeq_Jinfen" = c(0,6),
            "scRNASeq_BBN_C57B6" = c(0,22),
            "scRNASeq_BBN_Rag" = c(3),
            "scRNASeq_Chen" = c(15),
            "scRNASeq_Simon" = c())

l14 <- list("scRNASeq_Jinfen" = c(),
            "scRNASeq_BBN_C57B6" = c(14),
            "scRNASeq_BBN_Rag" = c(19),
            "scRNASeq_Chen" = c(7,26),
            "scRNASeq_Simon" = c(3,5,22))

l15 <- list("scRNASeq_Jinfen" = c(),
            "scRNASeq_BBN_C57B6" = c(6,9),
            "scRNASeq_BBN_Rag" = c(14),
            "scRNASeq_Chen" = c(),
            "scRNASeq_Simon" = c())

l16 <- list("scRNASeq_Jinfen" = c(13),
            "scRNASeq_BBN_C57B6" = c(1),
            "scRNASeq_BBN_Rag" = c(1),
            "scRNASeq_Chen" = c(8),
            "scRNASeq_Simon" = c())

l17 <- list("scRNASeq_Jinfen" = c(),
            "scRNASeq_BBN_C57B6" = c(),
            "scRNASeq_BBN_Rag" = c(28),
            "scRNASeq_Chen" = c(),
            "scRNASeq_Simon" = c(8,14))

# Create a dumym list to store markers from standard approach
marker_list <- list()
final_g <- tolower(unique(df_standard$gene))

for (list in c("l1", "l2", "l3", "l4", "l5", "l6", "l7", "l8", "l9", "l10", 
               "l11", "l12", "l13", "l14", "l15", "l16", "l17")){
  
  final_g <- tolower(unique(df_standard$gene))
  for (i in 1:length(get(list))){
    
    genes <- df_standard %>% 
      dplyr::filter(cluster %in% paste0(names(get(list)[i]),"_",  get(list)[[i]])) %>%
      dplyr::select(gene) %>%
      unlist(use.names = FALSE) %>%
      tolower()
    
    if(length(genes) > 0){
      final_g <- intersect(final_g, genes)
    }
  }
  marker_list <- c(marker_list, list(final_g))
}

max_l <- max(lengths(marker_list))
marker_list <- lapply(X=marker_list, FUN=function(x) { c(x, rep(x="", times=max_l-length(x)))})
df_final_standard <- data.frame(marker_list)

colnames(df_final_standard) <- c("Epithelial_1", "Epithelial_2", "Lymphatic.Endothelial", 
                  "Endothelial", "Granulocytes", "Plasma Cells", "Mitotic Cells", "B.Cells", "T.Cells",
                  "NK cells", "Macrophages", "Dendritic.Cells", "Mast Cells", 
                  "Myofibroblasts", "Fibroblasts_1", "Fibroblasts_2", 
                  "Fibroblasts_3")

# For custom approach
l1 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(8),
           "scRNASeq_BBN_Rag" = c(7),
           "scRNASeq_Chen" = c(7,19),
           "scRNASeq_Simon" = c(2,5,11))

l2 <- list("scRNASeq_Jinfen" = c(11),
           "scRNASeq_BBN_C57B6" = c(4,12),
           "scRNASeq_BBN_Rag" = c(4,6),
           "scRNASeq_Chen" = c(8),
           "scRNASeq_Simon" = c(9))

l3 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(11),
           "scRNASeq_BBN_Rag" = c(8),
           "scRNASeq_Chen" = c(),
           "scRNASeq_Simon" = c())

l4 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(6),
           "scRNASeq_BBN_Rag" = c(3),
           "scRNASeq_Chen" = c(5,6,15),
           "scRNASeq_Simon" = c(8))

l5 <- list("scRNASeq_Jinfen" = c(6,10),
           "scRNASeq_BBN_C57B6" = c(3),
           "scRNASeq_BBN_Rag" = c(5),
           "scRNASeq_Chen" = c(3),
           "scRNASeq_Simon" = c(7))

l6 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(5),
           "scRNASeq_BBN_Rag" = c(1),
           "scRNASeq_Chen" = c(1,16,18),
           "scRNASeq_Simon" = c(1,4))

l7 <- list("scRNASeq_Jinfen" = c(1),
           "scRNASeq_BBN_C57B6" = c(1),
           "scRNASeq_BBN_Rag" = c(2),
           "scRNASeq_Chen" = c(),
           "scRNASeq_Simon" = c())

l8 <- list("scRNASeq_Jinfen" = c(),
           "scRNASeq_BBN_C57B6" = c(),
           "scRNASeq_BBN_Rag" = c(),
           "scRNASeq_Chen" = c(11),
           "scRNASeq_Simon" = c(10))

l9 <- list("scRNASeq_Jinfen" = c(12),
            "scRNASeq_BBN_C57B6" = c(7),
            "scRNASeq_BBN_Rag" = c(),
            "scRNASeq_Chen" = c(10),
            "scRNASeq_Simon" = c())

l10 <- list("scRNASeq_Jinfen" = c(3,13),
            "scRNASeq_BBN_C57B6" = c(2),
            "scRNASeq_BBN_Rag" = c(2,17),
            "scRNASeq_Chen" = c(6),
            "scRNASeq_Simon" = c())

l11 <- list("scRNASeq_Jinfen" = c(9,14),
            "scRNASeq_BBN_C57B6" = c(10),
            "scRNASeq_BBN_Rag" = c(),
            "scRNASeq_Chen" = c(),
            "scRNASeq_Simon" = c())

# Create a dummy list to store markers from custom approach
marker_list <- list()
final_g <- tolower(unique(df_custom$gene))

for (list in c("l1", "l2", "l3", "l4", "l5", "l6", "l7", "l8", "l9", "l10", 
               "l11")){
  
  final_g <- tolower(unique(df_custom$gene))
  for (i in 1:length(get(list))){
    
    genes <- df_custom %>% 
      dplyr::filter(cluster %in% paste0(names(get(list)[i]),"_cluster_",  get(list)[[i]])) %>%
      dplyr::select(gene) %>%
      unlist(use.names = FALSE) %>%
      tolower()
    
    if(length(genes) > 0){
      final_g <- intersect(final_g, genes)
    }
  }
  marker_list <- c(marker_list, list(final_g))
}

max_l <- max(lengths(marker_list))
marker_list <- lapply(X=marker_list, FUN=function(x) { c(x, rep(x="", times=max_l-length(x)))})
df_final_custom <- data.frame(marker_list)

wb <- openxlsx::createWorkbook()
openxlsx::addWorksheet(wb = wb, sheetName = "Standard_Markers")
openxlsx::writeData(wb = wb, sheet = "Standard_Markers", x = df_final_standard)
openxlsx::addWorksheet(wb = wb, sheetName = "Custom_Markers")
openxlsx::writeData(wb = wb, sheet = "Custom_Markers", x = df_final_custom)
openxlsx::saveWorkbook(wb = wb, file = "Evaluated_Markers.xlsx", overwrite = TRUE)

#******************************************************************************#
#      STEP 5: VALIDATE THE SPECIFICITY FOR EACH MARKER GENE USING UMAPS       #
#******************************************************************************#

for (proj in c("scRNASeq_BBN_C57B6", "scRNASeq_BBN_Rag", "scRNASeq_Jinfen", 
               "scRNASeq_Chen", "scRNASeq_Simon")){
  
  seurat_results <- paste0("/hpc/home/kailasamms/scratch/", proj, "/results_seurat/")
  res <- 0.4
  reduc <- "rpca" #"harmony"
  celltype <- NULL
  #genes <- read.xlsx("/hpc/home/kailasamms/Markers.xlsx")
  genes <- read.xlsx("/hpc/home/kailasamms/projects/scRNASeq/scRNASeq_Markers.xlsx")
  
  # Load the integrated seurat object
  integrated_seurat <- readRDS(paste0(seurat_results, "integrated_seurat_snn", 
                                      dplyr::if_else(is.null(celltype), ".rds", paste0("_", celltype, ".rds"))))
  
  # Change active.ident to major clusters we annotated
  DefaultAssay(integrated_seurat) <- "RNA"
  idents <- paste0("cluster.", res, ".", base::tolower(reduc))
  Idents(object = integrated_seurat) <- idents
  
  feature_plot <- function(i){
    
    split <- "Condition" #NULL
    split <- NULL
    Seurat::FeaturePlot(object = integrated_seurat,
                        slot = "data",
                        features = i,
                        split.by = split,
                        cols =  c("#D1E5F0", "#F7F7F7", "#67001F"),
                        pt.size = 0.4,
                        order = TRUE,
                        min.cutoff = 'q10',
                        reduction = paste0("umap.", base::tolower(reduc)),
                        label = TRUE,
                        combine = TRUE,
                        raster = FALSE) +
      scale_colour_gradientn(colours = c("#D1E5F0", "#F7F7F7", "#67001F"), limits=c(0,2))
    # rev(brewer.pal(n = 11, name = "RdBu"))[5:11]
  } 
  
  for (i in 1:ncol(genes)){
    features <- genes[,i] %>% unlist(use.names=FALSE)
    features <- sort(rownames(integrated_seurat@assays$RNA$data)[tolower(rownames(integrated_seurat@assays$RNA$data)) %in% tolower(features)])
    
    # Plot UMAPs
    for (n in 1:ceiling((length(features)/12))){
      j <- 12*n-11
      k <- 12*n
      vec <- features[j:k]
      purrr::map(.x = vec[!is.na(vec)], 
                 .f = feature_plot) %>%
        cowplot::plot_grid(plotlist = .,
                           align = "hv",
                           axis = "tblr",
                           nrow = 3,
                           ncol = 4,
                           rel_widths = 1,
                           rel_heights = 1,
                           greedy = TRUE,
                           byrow = TRUE)
      
      ggplot2::ggsave(filename = paste0(proj,"_", colnames(genes)[i], "_", n, ".jpg"),
                      plot = last_plot(),
                      device = "jpeg",
                      scale = 1,
                      width = dplyr::if_else(is.null(split), 9*3, 9*6),
                      height = 11,
                      units = c("in"),
                      dpi = 300,
                      limitsize = FALSE,
                      bg = "white")
    }
  }
}


#******************************************************************************#

# # Find out clusters that are similar and genes that are common
# combinations_standard <- utils::combn(x=unique(df_standard$cluster), m=2)
# combinations_custom <- utils::combn(x=unique(df_custom$cluster), m=2)
# final_standard <- list()
# final_custom <- list()
# 
# for (i in 1:ncol(combinations_standard)){
#   
#   l <- intersect(df_standard %>% 
#                    dplyr::filter(cluster == combinations_standard[1,i]) %>%
#                    ungroup() %>%
#                    dplyr::select(gene) %>%
#                    unlist(use.names=FALSE) %>%
#                    tolower(),
#                  df_standard %>% 
#                    dplyr::filter(cluster == combinations_standard[2,i]) %>%
#                    ungroup() %>%
#                    dplyr::select(gene) %>%
#                    unlist(use.names=FALSE) %>%
#                    tolower())
#   
#   # Custom cutoff of 20 genes to be similar between 2 clusters
#   if (length(l) > 20){
#     l <-  list(l)
#     names(l) <- paste0(combinations_standard[1,i],".", combinations_standard[2,i])
#     final_standard <- c(final_standard, l)
#   }
# }
# 
# for (i in 1:ncol(combinations_custom)){
#   
#   l <- intersect(df_custom %>% 
#                    dplyr::filter(cluster == combinations_custom[1,i]) %>%
#                    ungroup() %>%
#                    dplyr::select(gene) %>%
#                    unlist(use.names=FALSE) %>%
#                    tolower(),
#                  df_custom %>% 
#                    dplyr::filter(cluster == combinations_custom[2,i]) %>%
#                    ungroup() %>%
#                    dplyr::select(gene) %>%
#                    unlist(use.names=FALSE) %>%
#                    tolower())
#   
#   # Custom cutoff of 20 genes to be similar between 2 clusters
#   if (length(l) > 20){
#     l <-  list(l)
#     names(l) <- paste0(combinations_custom[1,i],".", combinations_custom[2,i])
#     final_custom <- c(final_custom, l)
#   }
# }
# 
# # Convert list to dataframe
# for (i in 1:length(final_standard)){
#   n <- names(final_standard[i])
#   final_standard[i] <- list(c(unlist(final_standard[i], use.names=FALSE), rep(x="", times=max(lengths(final_standard))-length(final_standard[[i]]))))
#   names(final_standard[i]) <- names(final_standard[i])
# }
# 
# for (i in 1:length(final_custom)){
#   n <- names(final_custom[i])
#   final_custom[i] <- list(c(unlist(final_custom[i], use.names=FALSE), rep(x="", times=max(lengths(final_custom))-length(final_custom[[i]]))))
#   names(final_custom[i]) <- names(final_custom[i])
# }
# 
# common_markers_standard <- data.frame(final_standard)
# common_markers_custom <- data.frame(final_custom)
# 
# # Save all the markers and correlation
# wb <- openxlsx::createWorkbook()
# openxlsx::addWorksheet(wb = wb, sheetName = "Standard_Markers")
# openxlsx::writeData(wb = wb, sheet = "Standard_Markers", x = common_markers_standard )
# openxlsx::addWorksheet(wb = wb, sheetName = "Custom_Markers")
# openxlsx::writeData(wb = wb, sheet = "Custom_Markers", x = common_markers_custom)
# openxlsx::saveWorkbook(wb = wb, file = "Compiled_Markers.xlsx", overwrite = TRUE)